projects/dpg/product_fe_space_factory.h

```diff
@@ -0,0 +1,300 @@
+#ifndef PROJECTS_DPG_PRODUCT_FE_SPACE_FACTORY
+#define PROJECTS_DPG_PRODUCT_FE_SPACE_FACTORY
+
+/**
+ * @file
+ * @brief Factory class to create ProductUniformFEspaces.
+ * @author Philippe Peter
+ * @date June 2019
+ * @copyright MIT License
+ */
+
+#include <lf/base/base.h>
+#include <lf/mesh/mesh.h>
+#include <lf/uscalfe/uscalfe.h>
+
+#include "product_fe_space.h"
+
+#include "discontinuous_fe_constant.h"
+#include "lagr_fe_cubic.h"
+#include "lagr_fe_quadratic.h"
+
+#include "discontinuous_scalar_reference_finite_element.h"
+#include "dpg.h"
+#include "trace_scalar_reference_finite_element.h"
+
+namespace projects::dpg {
+
+/**
+ * @brief Factory class to build  a ProductUniformFESpace.
+ *
+ * This class can be used to construct a  ProductUniformFESpace. The class
+ * provides several methods to add new components  based on different continuity
+ * properties as well as the degree of the local shape functiosn .
+ */
+template <typename SCALAR>
+class ProductUniformFESpaceFactory {
```

Comment:
I think regarding its interface, this class is quite similar to the 
`MeshFactory` interface/classes. Also, as far as I have seen, all other builders in the lehrfempp library currently allow cascading function calls. This is however not possible with this class, since the methods should/have to return the index of the  added components. I agree that it  makes sense to rename the other two factory classes  into `ProductElementMatrixProviderBuilder` and `ProductElementVectorProviderBuilder`. I will also adapt their interfaces to allow cascading function calls in order to be more in line with the  lehrfempp builders I have seen so far.

Reply:
Yeah this is sort of a corner case. The distinction between `Factory` and `Builder` is very subtle but in a purely object oriented code, a (Abstract)Factory would really be (in my opinion) an interface class with a method to create objects + a set of implementations that create the different types of objects (which all inherit from the same base classs). In this sense the naming of `MeshFactory` is also a bit misleading but it originates from the Dune Mesh interface ;)

Lets keep ProductUniformFESpaceFactory then :)